pkg/workflow/workflow.go

```diff
@@ -0,0 +1,232 @@
+}
+
+type Option func(*Workflow)
+
+func WithDetails(details ...interface{}) Option {
+	return func(w *Workflow) {
+		w.logger = Logger{w.logger.WithValues(details...)}
+	}
+}
+
+func GetWorkflowFromObj(ctx context.Context, obj k8sRuntime.Object, details ...interface{}) (context.Context, Workflow) {
+	requestID := GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	var startTime time.Time
+	if startTimeStr != "" {
+		startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	w := Workflow{
+		requestID: requestID,
+		startTime: startTime,
+		logger:    Logger{klogr.New().WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, requestID)...)},
+	}
+	w.parentWorkflow = &w
+	return w.SaveToContext(ctx), w
+}
+
+func GetObjectDetails(obj k8sRuntime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+// New starts a new child workflow if there is already a workflow saved to the context otherwise a new workflow
+func New(ctx context.Context, opts ...Option) (context.Context, Workflow) {
+	name := getCallerName()
+
+	w := emptyWorkflow()
+	w.name = name
+
+	if parentWorkflow, ok := GetWorkflowFromContext(ctx); ok {
+		w.requestID = parentWorkflow.requestID
+		w.parentWorkflow = &parentWorkflow
+		w.logger = parentWorkflow.Logger()
+		_ = atomic.AddInt32(&w.parentWorkflow.numWait, 1)
+	} else {
+		w.requestID = GenerateRequestID()
+		w.logger = Logger{klogr.New().WithValues(consts.RequestIDKey, w.requestID)}
+	}
+
+	for _, opt := range opts {
+		opt(&w)
+	}
+
+	// so workflow's numWait will always be number of children workflow + 1 (itself)
+	return w.SaveToContext(ctx), w
+}
+
+func emptyWorkflow() Workflow {
+	w := Workflow{
+		startTime: time.Now(),
+		logger:    Logger{klogr.New()},
+		numWait:   1,
+		errSet:    &sync.Map{},
+	}
+	w.parentWorkflow = &w
+	return w
+}
+
+func (w Workflow) Finish(errs ...error) {
+	if len(errs) > 0 {
+		for _, err := range errs {
+			w.errSet.Store(err, nil)
+		}
+	}
+	if w.numWait > 0 {
+		return
+	}
+
+	latency := time.Since(w.startTime)
+	errs = w.errors()
+	if len(errs) == 0 {
+		err := status.Errorf(codes.Internal, "%+v", errs)
+		w.logger.WithValues(consts.Latency, latency, "workflow", w.name).WithCallDepth(1).Error(err, "Workflow completed with an error.")
+	} else {
+		w.logger.WithValues(consts.Latency, latency, "workflow", w.name).WithCallDepth(1).Info("Workflow completed with success.")
+	}
+
+	if w.parentWorkflow != nil {
+		_ = atomic.AddInt32(&w.parentWorkflow.numWait, -1)
```

Comment:
I believe the decrement should happen in the `Finish` method itself.

Reply:
Makes sense.